src/chardet.cpp

```diff
@@ -114,15 +114,20 @@ CHARDET_API short detect_handledata_r (Detect ** det, const char * buf, size_t b
 
 	ret = (*det)->detect->getCharsetName ();
 
-	if ( ! ret )
+	if ( ! ret ){
+		delete det;
```

Comment:
The det variable used by the detect_handledata_r function was generated by the detect_init function. Therefore, depending on the return value of detect_Handledata_r, it must be deleted by the detect_destroy function outside the detect_handledata_r.

Therefore, the detect_dataHandle_r function modification is not required as my thought.

For example, you must be processed as follows:

```c
d = detect_init ();
detect_reset (&d);

obj = detect_obj_init ();
swith (detect_handledata_r (&d, strings, string_length, &obj)) {
    case CHARDET_OUT_OF_MEMORY :
    case CHARDET_NULL_OBJECT :
        detect_obj_free (&obj);
        detect_destroy (&d);
        exit (1);
}
detect_obj_free (&obj);
detect_destroy (&d);
```

The detect_handledata_r function is provided for use in the loop. Therefore, the example of the following example is better:

```c
d = detect_init ();
if ( !d ) exit (1);

while ( condition ) {
    detect_reset (&d);
    obj = detect_obj_init ();

    swith (detect_handledata_r (&d, strings, string_length, &obj)) {
        case CHARDET_OUT_OF_MEMORY :
        case CHARDET_NULL_OBJECT :
            detect_obj_free (&obj);
            continue;
    }

    printf (
        "encoding: %s, confidence: %f, exist BOM: %d\n",
        obj->encoding, obj->confidence, obj->bom
    );
    detect_obj_free (&obj);
}
detect_destroy (&d);
```